### PR TITLE
Configure S3

### PR DESCRIPTION
**Pull request by jasonaowen**
There are a number of S3-compatible providers, and we should support them if it's reasonably easy to do so. Configuring the service to use a particular S3 provider involves several settings:

- the endpoint URL
- the region
- whether the provider uses subdomains or paths for referencing buckets
- the access key ID
- the access key secret
- the bucket

In #567, we were still thinking of having separate buckets for processed and unprocessed data. Later, in #564, we decided to only have one bucket. Consolidate and rename the bucket environment variables.

Finally, not having any of these environment variables set is a configuration error, and the app should not start in such an error state. Require the environment variables, and log an error & quit if they are missing.

Note that we should not deploy this until the relevant environment variables are set. I believe I have access and know how to do that in our current Docker setup.

### CODE REVIEW

File: src/s3Client.ts
Comment (slifty) [inline src/s3Client.ts:2]: This triggered a series of thoughts, so I'm writing them down:

1. I bet there are other places we should be more explicitly requiring env vars (doesn't have to be in this PR)
2. What does dotenv get us / should we still use dotenv (I think dotenv is providing the ability to autoload a `.dotenv` file in the event that env vars aren't automatically sourced, yes?  Do we care enough about that to warrant having two env var dependencies?)
3. I suppose a way to accomplish this within dotenv would be to create a class dedicated to environment variable management, and access env vars from that class.

None of these thoughts warrant a change to this PR necessarily, but I wanted to raise them.

I am comfortable with this approach as written, though I could imagine there being a benefit from a custom class that defines (and validates) environment variables even more robustly.

File: src/s3Client.ts
Comment (slifty) [inline src/s3Client.ts:10]: For my edification: this pattern will break immediately on process start if the vars aren't set / we don't need to also explicitly check for these in `src/index.ts` right?

File: .env.example
Comment (slifty) [inline .env.example:30]: This format doesn't work with `envsubst`, right? (E.g. if someone runs `envsubst .env.example > .env` I don't think they would be able to populate `S3_PATH_STYLE` without additional commands)

Can we maintain the `${S3_PATH_STYLE}` approach but include legitimate values in a comment?

**Review by slifty — CHANGES_REQUESTED** [review]
Much cleaner and safer thank you

One small request around the `.env.example` but otherwise ready to roll!

File: src/s3Client.ts
Comment (jasonaowen) [reply src/s3Client.ts:2]: > I bet there are other places we should be more explicitly requiring env vars

I actually think this is the last of them! `git grep process.env` is a pleasantly short and conditional.

> What does dotenv get us / should we still use dotenv

It is a convenience for developers, to automatically load the contents of a `.env` file. I am increasingly of the opinion that such convenience support should not live in the repo, but rather be the responsibility of the devs, via [direnv](https://direnv.net/) or another way of managing environment variables. However, I do not believe so strongly enough to advocate for removing existing, working setups, so for now my answer to the second question is "eh, sure".

> create a class dedicated to environment variable management

We could. I think the main benefit would be consolidating all the environment variables in one place; currently, if say `OPENAPI_DOCS_AUTH_CLIENT_ID` and `S3_ACCESS_KEY_ID` were both missing, only one would be reported because they're in separate calls to `requireEnv`. I'm not sure it's worth the work and conceptual overhead, but could probably be convinced.

File: src/s3Client.ts
Comment (jasonaowen) [reply src/s3Client.ts:10]: Yes, it will break on process start if any of the variables are not set, and no we do not need to also check in `src/index.ts`.

File: .env.example
Comment (jasonaowen) [reply .env.example:30]: Ahh, is that the motivation for this format? I wasn't sure why they were formatted using shell variables rather than example values. ...are we actually doing that anywhere? Should we document that? I'm not sure I understand the use case, tbh!

Comment (slifty) [reply .env.example]: It's just an accessibility thing that I've found to be best practice.  If you have a strong reason against it I'm happy to hear it!  I think signaling values can easily be done through comments (and prefer that anyway since the comments live on once the value is replaced)

File: src/s3Client.ts
Comment (slifty) [reply src/s3Client.ts:2]: > I actually think this is the last of them!

I *may* be missing it but it looks like this doesn't get at the env var in https://github.com/PhilanthropyDataCommons/service/blob/main/src/logger.ts right?

> It is a convenience for developers

Ahhhh -- then... I do think we should plan to get rid of it especially due to the awkwardness documented in `index.ts` around import orders mattering (but I'll open an issue for it).

(the class manager convo I'm happy to let sit as a musing in a PR thread and not carry it forward at all!)

**Review by slifty — APPROVED** [review]
looks good!  There's one instance of a process.env being directly accessed in another file but that isn't related to S3.  I defer to you on if you wanna include it in here.

File: src/s3Client.ts
Comment (slifty) [reply src/s3Client.ts:2]: (as noted on call -- this is not required so requiring wouldn't be good lulz)

Comment (jasonaowen) [reply src/s3Client.ts:2]: Captured in https://github.com/PhilanthropyDataCommons/service/issues/640!

File: .env.example
Comment (jasonaowen) [reply .env.example]: Thanks for explaining!

**Comment by jasonaowen** [comment]
Rebased onto `main`; will merge once CI passes.

**Pull request merged**
